### ISSUE TITLE
Is it possible to implement predict_proba?

### ISSUE THREAD

**Issue by thiagobrito** [trevorstephens/gplearn#24]
I don't know if it's possible to implement, but it could be really important in some kind of solutions!

**Comment by trevorstephens** [comment]
There is currently no classifier in this package so probs aren't really applicable. Probably aiming for 0.3 release at this point to add a classifier. Probabilities might be tough for this though as it is not an ensemble method (like RF/GBM which get probs from voting) and cross-validation is tough in GP (that's how SVMs do probs). I could expose the decision function perhaps, but it is possibly a stretch to say that distance from the decision boundary is the same as a probability. Shall keep the issue open as a potential feature.

**Comment by MaxHalford** [comment]
It seems to be used [here](https://www.kaggle.com/scirpus/genetic-programming-lb-0-88).

**Comment by guyko81** [comment]
Hi Trevor,

As far as I see there's no classifier yet, right? 
The solution in my head is the following: 
1) build as many regressions as classes are
2) use the categorical crossentropy with logits as a metric in the concatenation of the outputs
3) handle the different labels' regressions separately

this is a code for categorical crossentropy written in numpy, the labels are integers starting with values between 0 and num_of_classes-1, logits are simple numbers between minus infinity and infinity
logits shape: (N, C), labels shape: (N,)
N is the number of rows in the dataset, C is the number of labels
logits is a concatenation of C separate (N,) shape vectors

def loss(logits, labels):
        logits = logits - np.max(logits, axis=-1, keepdims=True)
        logits_at_labels = logits[np.arange(len(logits)), labels]
        crossentropy = -logits_at_labels + np.log(np.sum(np.exp(logits), axis=-1))
        return crossentropy

**Comment by trevorstephens** [comment]
@guyko81 I've started working on a binary classifier and hope to release a new version of gplearn with this included in the next month or so. Keep an eye on #141 which is the classifier feature branch (very early so far! Bit more locally not yet pushed). Currently no plan to support multi-label classification.

**Comment by guyko81** [comment]
in that case the solution could be to simply change the outcome (compared to the normal regression model) with the written sigmoid function to be between 0 and 1, and use the log_loss from sklearn.metrics:

def _binary_crossentropy_loss(y, y_pred, w):
    """Calculate the categorical crossentropy error."""
    return log_loss(y, y_pred, sample_weight=w)

**Comment by trevorstephens** [comment]
Yeah that's the idea
